ugvc/mrd/bqsr_inference_utils.py

```diff
@@ -0,0 +1,205 @@
+    # extract motifs and determine cskp (usually done in featuremap_to_dataframe)
+    df = (
+        get_motif_around(df.assign(indel=False), motif_size=4, fasta=params["ref_fasta"])
+        .drop(columns=["indel"])
+        .astype({"left_motif": str, "right_motif": str})
+    )
+    left_motif_reverse = df["left_motif"].apply(revcomp)
+    right_motif_reverse = df["right_motif"].apply(revcomp)
+    is_reverse = (df["X_FLAGS"] & 16).astype(bool)
+
+    # Reverse complement the ref and alt alleles if necessary to match the read direction
+    for column in ("ref", "alt"):
+        df[column] = df[column].where(is_reverse, df[column].apply(revcomp))
+    df.loc[:, "left_motif"] = df["left_motif"].where(is_reverse, right_motif_reverse)
+    df.loc[:, "right_motif"] = df["right_motif"].where(is_reverse, left_motif_reverse)
+
+    df.loc[:, "ref_motif"] = df["left_motif"].str.slice(-1) + df["ref"] + df["right_motif"].str.slice(0, 1)
+    df.loc[:, "alt_motif"] = df["left_motif"].str.slice(-1) + df["alt"] + df["right_motif"].str.slice(0, 1)
+    flow_order = "TGCA"
+    df_cskp = get_cycle_skip_dataframe(flow_order=flow_order)
+    df = df.set_index(["ref_motif", "alt_motif"]).join(df_cskp).reset_index()
+
+    df = df.assign(is_cycle_skip=df["cycle_skip_status"] == "cycle-skip")
+
+    df.loc[:, "ref_alt_bases"] = df["ref"].str.cat(df["alt"], sep="")
+    df.loc[:, "ref_alt_motif"] = df["ref_motif"].str.cat(df["alt"], sep="")
+    df.loc[:, "left_motif"] = df["left_motif"].str.slice(start=0, stop=3)
+    df.loc[:, "right_motif"] = df["right_motif"].str.slice(start=1, stop=4)
+
+    # # ['cycle_skip_status', 'ref_alt_motif', 'left_motif', 'right_motif',
+    # #   'X_SCORE', 'X_EDIST', 'X_LENGTH', 'X_MAPQ', 'X_INDEX', 'is_mixed']
+
+    features = df[model.feature_names_in_]
+    motif_only_features = ["ref_alt_motif", "left_motif", "right_motif"]
+    categorical_features = ["cycle_skip_status"] + motif_only_features
+
+    features = features.astype({col: "category" for col in categorical_features})
+
+    # scaled_features = preprocessing.scale(features)  # Scale features if needed
+    predicted_qualities = model.predict_proba(features)
+
+    # Assign the new quality value to each variant and write to the output file
+    for i, variant in enumerate(chunk):
+        # Add a new INFO field to the variant indicating the ML quality value
+        # variant.info.__setitem__("ML_QUAL", -10 * np.log10(predicted_qualities[i][0]))
+        variant.info["ML_QUAL"] = -10 * np.log10(predicted_qualities[i][0])
+        # Write the variant with the new quality to the output VCF file
+        vcf_out.write(variant)
+
+
+def bqsr_inference(
+    featuremap_path,
+    params_path,
+    model_path,
+    out_path: str,
+    out_basename: str = "",
+):
+
+    # init
+    os.makedirs(out_path, exist_ok=True)
+    if len(out_basename) > 0 and not out_basename.endswith("."):
+        out_basename += "."
+
+    params = {}
```

Comment:
I don't understand the purpose of this dictionary, I would either keep these as separate variables or create a class like Doron suggested.